rust/lance-file/src/reader.rs

```diff
@@ -0,0 +1,1800 @@
+    let span = tracing::span!(tracing::Level::DEBUG, "apply_deletions");
+    let _enter = span.enter();
+    let deletion_mask =
+        deletion_vector.and_then(|v| v.build_predicate(row_ids.as_ref().unwrap().iter()));
+
+    let batch = match batch {
+        Some(batch) => {
+            if with_row_id {
+                let row_id_arr = Arc::new(UInt64Array::from(row_ids.unwrap()));
+                batch.try_with_column(ROW_ID_FIELD.clone(), row_id_arr)?
+            } else {
+                batch
+            }
+        }
+        None if with_row_id => {
+            let row_id_arr = Arc::new(UInt64Array::from(row_ids.unwrap()));
+            RecordBatch::try_new(
+                Arc::new(ArrowSchema::new(vec![ROW_ID_FIELD.clone()])),
+                vec![row_id_arr as ArrayRef],
+            )?
+        }
+        _ => unreachable!(),
+    };
+
+    match (deletion_mask, reader.make_deletions_null) {
+        (None, _) => Ok(batch),
+        (Some(mask), false) => Ok(filter_record_batch(&batch, &mask)?),
+        (Some(mask), true) => Ok(apply_deletions_as_nulls(batch, &mask)?),
+    }
+}
+
+/// Apply a mask to the batch, where rows are "deleted" by the _rowid column null.
+/// This is used as a performance optimization to avoid copying data.
+fn apply_deletions_as_nulls(batch: RecordBatch, mask: &BooleanArray) -> Result<RecordBatch> {
+    // Transform mask into null buffer. Null means deleted, though note that
+    // null buffers are actually validity buffers, so True means not null
+    // and thus not deleted.
+    let mask_buffer = NullBuffer::new(mask.values().clone());
+
+    match mask_buffer.null_count() {
+        // All rows are deleted
+        n if n == mask_buffer.len() => return Ok(RecordBatch::new_empty(batch.schema())),
+        // No rows are deleted
+        0 => return Ok(batch),
+        _ => {}
+    }
+
+    // For each column convert to data
+    let new_columns = batch
+        .schema()
+        .fields()
+        .iter()
+        .zip(batch.columns())
+        .map(|(field, col)| {
+            if field.name() == ROW_ID {
+                let col_data = col.to_data();
+                // If it already has a validity bitmap, then AND it with the mask.
+                // Otherwise, use the boolean buffer as the mask.
+                let null_buffer = NullBuffer::union(col_data.nulls(), Some(&mask_buffer));
+
+                Ok(col_data
+                    .into_builder()
+                    .null_bit_buffer(null_buffer.map(|b| b.buffer().clone()))
+                    .build()
+                    .map(make_array)?)
+            } else {
+                Ok(col.clone())
+            }
+        })
+        .collect::<Result<Vec<_>>>()?;
+
+    Ok(RecordBatch::try_new(batch.schema(), new_columns)?)
+}
+
+#[async_recursion]
+async fn read_array(
+    reader: &FileReader,
+    field: &Field,
+    batch_id: i32,
+    page_table: &PageTable,
+    params: &ReadBatchParams,
+) -> Result<ArrayRef> {
```

Comment:
Why are these methods in this file? It seems like this belongs with the `FileReader` elsewhere?

Reply:
Are you saying this should be an `impl FileReader` method instead of a free function?  I think I agree but that change was not introduced in this file (it just looks like it because of the `lance-file` mishap)